devicedefender/source/DeviceDefender.cpp

```diff
@@ -22,11 +22,81 @@ namespace Aws
 
     namespace Iotdevicedefenderv1
     {
-        // Define statics
-        unsigned long long ReportTask::s_cpuLastTotalUser = 0;
-        unsigned long long ReportTask::s_cpuLastTotalUserLow = 0;
-        unsigned long long ReportTask::s_cpuLastTotalSystem = 0;
-        unsigned long long ReportTask::s_cpuLastTotalIdle = 0;
+        // Custom number metric setup and metric function getter.
+        void CustomMetricNumber::SetCustomMetricData(
+            CustomMetricNumberFunction inputFunction,
+            Crt::Allocator *inputAllocator)
+        {
+            function = std::move(inputFunction);
+            m_allocator = inputAllocator;
+        }
+        int CustomMetricNumber::GetMetricFunction(double *output, void *data)
+        {
+            CustomMetricNumber *metric = (CustomMetricNumber *)data;
+            return metric->function(output);
+        }
+
+        // Custom number list metric setup and metric function getter.
+        void CustomMetricNumberList::SetCustomMetricData(
+            CustomMetricNumberListFunction inputFunction,
+            Crt::Allocator *inputAllocator)
+        {
+            function = std::move(inputFunction);
+            m_allocator = inputAllocator;
+        }
+        int CustomMetricNumberList::GetMetricFunction(aws_array_list *output, void *data)
+        {
+            CustomMetricNumberList *metric = (CustomMetricNumberList *)data;
+            Crt::Vector<double> function_data = Crt::Vector<double>();
+            int returnValue = metric->function(&function_data);
+            for (size_t i = 0; i < function_data.size(); i++)
+            {
+                aws_array_list_push_back(output, &function_data.at(i));
+            }
+            return returnValue;
+        }
+
+        // Custom string list metric setup and metric function getter.
+        void CustomMetricStringList::SetCustomMetricData(
+            CustomMetricStringListFunction inputFunction,
+            Crt::Allocator *inputAllocator)
```

Comment:
It's not clear why inputAllocator is needed anywhere in these.

Reply:
It is needed for making an `aws_string` in the GetMetricFunction for `StringList`s and `IPAddressList`s.